### PR TITLE
[Merged by Bors] - feat: density of a finite kernel wrt another kernel

### PR DESCRIPTION
**Pull request by RemyDegenne**
Let `κ : kernel α (γ × β)` and `ν : kernel α γ` be two finite kernels with `kernel.fst κ ≤ ν`, where `γ` has a countably generated σ-algebra (true in particular for standard Borel spaces).
We build a function `f : α → γ → Set β → ℝ` jointly measurable in the first two arguments such that for all `a : α` and all measurable sets `s : Set β` and `A : Set γ`, `∫ x in A, f a x s ∂(ν a) = (κ a (A ×ˢ s)).toReal`.

---

- [x] depends on: #10945
- [x] depends on: #10946

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Probability/Kernel/Disintegration/Density.lean [hunk 1]
```diff
@@ -0,0 +1,799 @@
...
+
+lemma integrable_densityProcess (hκν : fst κ ≤ ν) [IsFiniteKernel ν] (n : ℕ)
+    (a : α) {s : Set β} (hs : MeasurableSet s) :
+    Integrable (fun x ↦ densityProcess κ ν n a x s) (ν a) := by
+  rw [← memℒp_one_iff_integrable]
+  refine ⟨Measurable.aestronglyMeasurable ?_, ?_⟩
+  · exact measurable_densityProcess_right κ ν n a hs
+  · exact (snorm_densityProcess_le hκν n a s).trans_lt (measure_lt_top _ _)
+
+lemma set_integral_densityProcess_I (hκν : fst κ ≤ ν) [IsFiniteKernel κ] [IsFiniteKernel ν]
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration/Density.lean:185]: I don't understand the naming `_I`. Also, doesn't the finiteness assumption on kappa follow from `hκν` (here and below)?

File: Mathlib/Probability/Kernel/Disintegration/Density.lean [hunk 2]
```diff
@@ -0,0 +1,799 @@
...
+  · exact (integrable_densityProcess hκν n a hs).integrableOn
+  rw [ENNReal.tsum_toReal_eq (fun _ ↦ measure_ne_top _ _)]
+  congr with k
+  rw [set_integral_densityProcess_I hκν _ _ hs k.prop]
+
+lemma set_integral_densityProcess (hκν : fst κ ≤ ν) [IsFiniteKernel κ] [IsFiniteKernel ν]
+    (n : ℕ) (a : α) {s : Set β} (hs : MeasurableSet s) {A : Set γ}
+    (hA : MeasurableSet[countableFiltration γ n] A) :
+    ∫ x in A, densityProcess κ ν n a x s ∂(ν a) = (κ a (A ×ˢ s)).toReal := by
+  refine induction_on_inter (m := countableFiltration γ n) (s := countablePartition γ n)
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration/Density.lean:252]: It seems a little bit far-fetched to use a pi-system argument in such a simple situation. Do you have the fact that any set measurable wrt `countableFiltration γ n` is a finite union of atoms? From this, the conclusion would be direct from `set_integral_densityProcess_I`, without the need for ` integral_densityProcess`.

File: Mathlib/Probability/Kernel/Disintegration/Density.lean [hunk 3]
```diff
@@ -0,0 +1,799 @@
...
+
+@[simp]
+lemma densityProcess_empty (κ : kernel α (γ × β)) (ν : kernel α γ) (n : ℕ) (a : α) (x : γ) :
+    densityProcess κ ν n a x ∅ = 0 := by
+  simp [densityProcess]
+
+lemma tendsto_densityProcess_atTop_empty_of_antitone (κ : kernel α (γ × β)) (ν : kernel α γ)
+    [IsFiniteKernel κ] (n : ℕ) (a : α) (x : γ)
+    (s : ℕ → Set β) (hs : Antitone s) (hs_iInter : ⋂ i, s i = ∅)
+    (hs_meas : ∀ n, MeasurableSet (s n)) :
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration/Density.lean:375]:
```suggestion
    (hs_meas : ∀ m, MeasurableSet (s m)) :
```
(keep `n` for the index of the countable partition). Same thing below several times.

File: Mathlib/Probability/Kernel/Disintegration/Density.lean [hunk 4]
```diff
@@ -0,0 +1,799 @@
...
+    ∀ᵐ x ∂(ν a), Tendsto (fun n ↦ densityProcess κ ν n a x s) atTop
+      (𝓝 ((countableFiltration γ).limitProcess
+      (fun n x ↦ densityProcess κ ν n a x s) (ν a) x)) := by
+  refine Submartingale.ae_tendsto_limitProcess (martingale_densityProcess hκν a hs).submartingale
+    (R := (ν a univ).toNNReal) (fun n ↦ ?_)
+  refine (snorm_densityProcess_le hκν n a s).trans_eq ?_
+  rw [ENNReal.coe_toNNReal]
+  exact measure_ne_top _ _
+
+lemma limitProcess_densityProcess_mem_L1 (hκν : fst κ ≤ ν) [IsFiniteKernel κ] [IsFiniteKernel ν]
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration/Density.lean:417]:
```suggestion
lemma memL1_limitProcess_densityProcess (hκν : fst κ ≤ ν) [IsFiniteKernel κ] [IsFiniteKernel ν]
```

File: Mathlib/Probability/Kernel/Disintegration/Density.lean [hunk 5]
```diff
@@ -0,0 +1,799 @@
...
+  · rw [integral_zero]
+    exact tendsto_integral_density_of_antitone hκν a s hs hs_iInter hs_meas
+  · exact ae_of_all _ (fun c n m hnm ↦ density_mono_set hκν a c (hs hnm))
+  · exact ae_of_all _ (fun x m ↦ density_nonneg hκν a x (s m))
+
+section UnivFst
+
+/-! We specialize to `ν = fst κ`, for which `density κ (fst κ) a t univ = 1` almost everywhere. -/
+
+lemma densityProcess_univ [IsFiniteKernel κ] (n : ℕ) (a : α) (x : γ) :
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration/Density.lean:679]: The naming of this lemma (and of the following ones) is a little bit too generic, as it does not indicate you are considering a specific situation. Add `fst` somewhere, here and below?

**Review by sgouezel — COMMENTED** [review]
Looks great, thanks!

File: Mathlib/Probability/Kernel/Disintegration/Density.lean [hunk 6]
```diff
@@ -0,0 +1,799 @@
...
+Authors: Rémy Degenne
+-/
+import Mathlib.Probability.Kernel.Composition
+import Mathlib.Probability.Martingale.Convergence
+import Mathlib.Probability.Process.PartitionFiltration
+
+/-!
+# Kernel density
+
+Let `κ : kernel α (γ × β)` and `ν : kernel α γ` be two finite kernels with `kernel.fst κ ≤ ν`,
```
[the remark this answers is not in the document]
Comment (RemyDegenne) [reply Mathlib/Probability/Kernel/Disintegration/Density.lean:13]: I have two applications of that code, one for which beta is important and one for which it isn't:
- for kernel disintegration, I want to take `κ : kernel α (γ × β)` and write it as a composition-product of `fst κ` and a `kernel (α × γ) β`. I will go through beta=R with sets `s` taken to be `Iic q` for rational `q`, and build a kernel by a kernel cdf construction, hence the monotone sequences of sets.
- for the Radon-Nikodym theorem for kernels I don't need beta and will drop it.

I'll add a paragraph about the uses of the `density` construction and explain that there.

File: Mathlib/Probability/Kernel/Disintegration/Density.lean [hunk 7]
```diff
@@ -0,0 +1,786 @@
...
+    rwa [ne_eq, ← Subtype.ext_iff]
+  rw [integral_iUnion]
+  · rw [iUnion_prod_const, measure_iUnion]
+    · rw [ENNReal.tsum_toReal_eq (fun _ ↦ measure_ne_top _ _)]
+      congr with u
+      rw [set_integral_densityProcess_of_mem hκν _ _ hs (hS_subset (by simp))]
+    · intro u v huv
+      simp only [Finset.coe_sort_coe, Set.disjoint_prod, disjoint_self, bot_eq_empty]
+      exact Or.inl (h_disj huv)
+    · exact fun _ ↦ (measurableSet_countablePartition n (hS_subset (by simp))).prod hs
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration/Density.lean:258]:
```suggestion
  rw [integral_iUnion, iUnion_prod_const, measure_iUnion,
      ENNReal.tsum_toReal_eq (fun _ ↦ measure_ne_top _ _)]
  · congr with u
    rw [set_integral_densityProcess_of_mem hκν _ _ hs (hS_subset (by simp))]
  · intro u v huv
    simp only [Finset.coe_sort_coe, Set.disjoint_prod, disjoint_self, bot_eq_empty]
    exact Or.inl (h_disj huv)
  · exact fun _ ↦ (measurableSet_countablePartition n (hS_subset (by simp))).prod hs
```
You can combine the `rw` to get one indentation level less.

**Review by sgouezel — COMMENTED** [review]
bors d+
Thanks!

**Pull request closed without merging**